### PR TITLE
Add support for Azure Workload Identity

### PR DESCRIPTION
**Pull request by nilsgstrabo**

### CODE REVIEW

**Review by satr — CHANGES_REQUESTED** [review]

File: pkg/apis/radix/v1/radixapptypes.go [hunk 1]
```diff
@@ -515,6 +519,17 @@ type OAuth2CookieStore struct {
...
 
+// Identity configuration for federation with external identity providers
+type Identity struct {
+	// Azure identity configuration
+	Azure *AzureIdentity `json:"azure,omitempty" yaml:"azure,omitempty"`
+}
+
+// AzureIdentity properties for Azure AD Workload Identity
+type AzureIdentity struct {
+	ClientId string `json:"clientId" yaml:"clientId"`
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/radix/v1/radixapptypes.go:532]: fixed

File: pkg/apis/utils/labels/labels.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+// ForCommitId returns labels describing the commit ID,
+// e.g. "radix-commit": "64b54f4a6aa542cb4bd15666c1e104eee647573a"
+func ForCommitId(commitId string) kubelabels.Set {
+	return kubelabels.Set{
+		kube.RadixCommitLabel: commitId,
+	}
+}
+
+// ForCommitId returns labels indicating that a pod is a job scheduler,
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/utils/labels/labels.go]: fixed

File: pkg/apis/deployment/radixcommoncomponent.go [hunk 3]
```diff
@@ -160,3 +161,39 @@ func getAzureKeyVaultItemsWithEnvVarsNotExistingInEnvSecretRefs(commonAzureKeyVa
...
+				err = e
+			} else {
+				panic(r)
+			}
+		}
+	}()
+
+	identity = &v1.Identity{}
+
+	if !reflect.ValueOf(radixComponent).IsNil() {
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/radixcommoncomponent.go]: fixed

File: pkg/apis/deployment/radixcommoncomponent.go [hunk 4]
```diff
@@ -160,3 +161,39 @@ func getAzureKeyVaultItemsWithEnvVarsNotExistingInEnvSecretRefs(commonAzureKeyVa
...
+		if componentIdentity := radixComponent.GetIdentity(); componentIdentity != nil {
+			componentIdentity.DeepCopyInto(identity)
+		}
+	}
+
+	if !reflect.ValueOf(environmentConfig).IsNil() {
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/radixcommoncomponent.go]: fixed

File: pkg/apis/deployment/serviceaccountspec.go [hunk 5]
```diff
@@ -46,20 +46,25 @@ func (spec *jobSchedulerServiceAccountSpec) AutomountServiceAccountToken() *bool
...
+type radixComponentServiceAccountSpec struct {
+	component v1.RadixCommonDeployComponent
+}
 
 func (spec *radixComponentServiceAccountSpec) ServiceAccountName() string {
-	return ""
+	if componentRequiresServiceAccount(spec.component) {
+		return utils.GetComponentServiceAccountName(spec.component.GetName())
+	}
+	return "default"
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/serviceaccountspec.go]: fixed

File: pkg/apis/deployment/serviceaccountspec_test.go [hunk 6]
```diff
@@ -16,17 +17,43 @@ func Test_ServiceAccountSpec(t *testing.T) {
...
+		assert.Equal(t, utils.BoolPtr(false), spec.AutomountServiceAccountToken())
+		assert.Equal(t, utils.GetComponentServiceAccountName(rd.Spec.Components[1].Name), spec.ServiceAccountName())
+
+		spec = NewServiceAccountSpec(rd, newJobSchedulerComponent(&rd.Spec.Jobs[0], rd))
+		assert.Equal(t, utils.BoolPtr(true), spec.AutomountServiceAccountToken())
+		assert.Equal(t, defaults.RadixJobSchedulerServerServiceName, spec.ServiceAccountName())
 
 		spec = NewServiceAccountSpec(rd, &rd.Spec.Jobs[0])
+		assert.Equal(t, utils.BoolPtr(false), spec.AutomountServiceAccountToken())
+		assert.Equal(t, "default", spec.ServiceAccountName())
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/serviceaccountspec_test.go]: fixed

File: pkg/apis/deployment/serviceaccountspec_test.go [hunk 7]
```diff
@@ -43,9 +70,14 @@ func Test_ServiceAccountSpec(t *testing.T) {
...
 		assert.Equal(t, defaults.RadixAPIServiceAccountName, spec.ServiceAccountName())
 
-		spec = NewServiceAccountSpec(rd, &rd.Spec.Jobs[0])
+		spec = NewServiceAccountSpec(rd, newJobSchedulerComponent(&rd.Spec.Jobs[0], rd))
 		assert.Equal(t, utils.BoolPtr(true), spec.AutomountServiceAccountToken())
 		assert.Equal(t, defaults.RadixJobSchedulerServerServiceName, spec.ServiceAccountName())
+
+		spec = NewServiceAccountSpec(rd, &rd.Spec.Jobs[0])
+		assert.Equal(t, utils.BoolPtr(false), spec.AutomountServiceAccountToken())
+		assert.Equal(t, "default", spec.ServiceAccountName())
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/serviceaccountspec_test.go]: fixed

File: pkg/apis/deployment/serviceaccountspec_test.go [hunk 8]
```diff
@@ -62,8 +94,13 @@ func Test_ServiceAccountSpec(t *testing.T) {
...
 		assert.Equal(t, defaults.RadixGithubWebhookServiceAccountName, spec.ServiceAccountName())
 
-		spec = NewServiceAccountSpec(rd, &rd.Spec.Jobs[0])
+		spec = NewServiceAccountSpec(rd, newJobSchedulerComponent(&rd.Spec.Jobs[0], rd))
 		assert.Equal(t, utils.BoolPtr(true), spec.AutomountServiceAccountToken())
 		assert.Equal(t, defaults.RadixJobSchedulerServerServiceName, spec.ServiceAccountName())
+
+		spec = NewServiceAccountSpec(rd, &rd.Spec.Jobs[0])
+		assert.Equal(t, utils.BoolPtr(false), spec.AutomountServiceAccountToken())
+		assert.Equal(t, "default", spec.ServiceAccountName())
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/deployment/serviceaccountspec_test.go]: fixed

File: pkg/apis/utils/labels/labels_test.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+func Test_ForPodWithRadixIdentity(t *testing.T) {
+	actual := ForPodWithRadixIdentity(nil)
+	assert.Equal(t, kubelabels.Set(nil), actual)
+
+	actual = ForPodWithRadixIdentity(&v1.Identity{})
+	assert.Equal(t, kubelabels.Set(nil), actual)
+
+	actual = ForPodWithRadixIdentity(&v1.Identity{Azure: &v1.AzureIdentity{ClientId: "any"}})
+	expected := kubelabels.Set{"azure.workload.identity/use": "true"}
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/utils/labels/labels_test.go:65]: The label key is defined by Azure Workload Identity and should be exactly this value. If the value of the const is changed, the test must fail

File: pkg/apis/utils/annotations/annotations_test.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+func Test_ForServiceAccountWithRadixIdentity(t *testing.T) {
+	actual := ForServiceAccountWithRadixIdentity(nil)
+	assert.Equal(t, map[string]string(nil), actual)
+
+	actual = ForServiceAccountWithRadixIdentity(&v1.Identity{})
+	assert.Equal(t, map[string]string(nil), actual)
+
+	actual = ForServiceAccountWithRadixIdentity(&v1.Identity{Azure: &v1.AzureIdentity{ClientId: "anyclientid"}})
+	expected := map[string]string{"azure.workload.identity/client-id": "anyclientid"}
```
[the remark this answers is not in the document]
Comment (nilsgstrabo) [reply pkg/apis/utils/annotations/annotations_test.go:46]: The annotation key is defined by Azure Workload Identity and should be exactly this value. If the value of the const is changed, the test must fail

**Review by satr — APPROVED** [review]

**Review by satr — APPROVED** [review]

**Pull request merged**